source/extensions/filters/network/redis_proxy/router_impl.h

```diff
@@ -70,11 +72,17 @@ class PrefixRoutes : public Router {
 
   RouteSharedPtr upstreamPool(std::string& key) override;
 
+  void setReadFilterCallback(Network::ReadFilterCallbacks* callbacks) override {
+    callbacks_ = callbacks;
+  }
+
 private:
   TrieLookupTable<PrefixSharedPtr> prefix_lookup_table_;
   const bool case_insensitive_;
   Upstreams upstreams_;
   RouteSharedPtr catch_all_route_;
+  Network::ReadFilterCallbacks* callbacks_{};
```

Comment:
I wonder whether we should use a shared pointer here so that we won't hit use-after-free issue.

Reply:
i'm not quite sure, I was trying to lookup for references in the existing codebase to get some inspiration but I couldn't find any. Alternately, would it be better to just do a `nullptr` check on the callbacks object before using it ?

Reply:
I'm a little worried that we could use the pointer while the object has been freed, causing Envoy crash.

We could use shared pointer, for example:
https://github.com/envoyproxy/envoy/blob/f179528d6fadc4f3357d59ee0e9ffa9307f4c8ee/envoy/event/dispatcher.h#L58

However, it looks like that the lifetime of object is not managed by Redis filter itself, which means that the object's lifetime should cover the lifetime of the Redis filter:
https://github.com/envoyproxy/envoy/blob/f179528d6fadc4f3357d59ee0e9ffa9307f4c8ee/source/extensions/filters/network/redis_proxy/proxy_filter.cc#L84

Sounds ok to use the raw pointer here.

Reply:
@weisisea - sorry I got a little confused, does that mean we are good over here or I should still explore using `shared_ptr` ? Thought of double checking with you once.

Reply:
Sorry for not being clear, we can leave the changes as they are, no need to explore shared pointer.